pkg/auth/providers/ldap/ldap_client.go

```diff
@@ -158,17 +159,101 @@ func (p *ldapProvider) getPrincipalsFromSearchResult(result *ldapv2.SearchResult
 		query := fmt.Sprintf("(&(%v=%v)(objectClass=%v))", config.GroupMemberMappingAttribute, ldapv2.EscapeFilter(groupMemberUserAttribute[0]), config.GroupObjectClass)
 		newGroupPrincipals, err := p.searchLdap(query, groupScope, config, lConn)
 		//deduplicate groupprincipals get from userMemberAttribute
-		nonDupGroupPrincipals = p.findNonDuplicateGroupPrincipals(newGroupPrincipals, groupPrincipals, nonDupGroupPrincipals)
+		nonDupGroupPrincipals = p.findNonDuplicateBetweenGroupPrincipals(newGroupPrincipals, groupPrincipals, nonDupGroupPrincipals)
 		groupPrincipals = append(groupPrincipals, nonDupGroupPrincipals...)
 		if err != nil {
 			return userPrincipal, groupPrincipals, err
 		}
 	}
 
+	searchDomain := config.UserSearchBase
+	if config.GroupSearchBase != "" {
+		searchDomain = config.GroupSearchBase
+	}
+
+	// Handling nestedgroups: tracing from down to top inorder to find the parent groups, parent parent groups, and so on...
+	// When traversing up, we note down all the parent groups and add them to groupPrincipals
+	for _, groupPrincipal := range groupPrincipals {
+		err := p.tracingUpToFindParentGroups(groupPrincipal, searchDomain, groupScope, config, lConn)
+		if err != nil {
+			return userPrincipal, groupPrincipals, nil
+		}
+	}
+	nestedGroupPrincipals = p.deDupWithInGroupPrincipals(nestedGroupPrincipals)
+	nonDupGroupPrincipals = p.findNonDuplicateBetweenGroupPrincipals(nestedGroupPrincipals, groupPrincipals, []v3.Principal{})
+	groupPrincipals = append(groupPrincipals, nonDupGroupPrincipals...)
+
 	return userPrincipal, groupPrincipals, nil
 }
 
-func (p *ldapProvider) findNonDuplicateGroupPrincipals(newGroupPrincipals []v3.Principal, groupPrincipals []v3.Principal, nonDupGroupPrincipals []v3.Principal) []v3.Principal {
+func (p *ldapProvider) tracingUpToFindParentGroups(groupPrincipal v3.Principal, searchDomain string, groupScope string, config *v3.LdapConfig, lConn *ldapv2.Conn) error {
```

Comment:
the name of this function is awkward. Can we change the name to `gatherParentGroups`?